subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/PropertyReportTask.java

```diff
@@ -60,23 +67,81 @@ public void setRenderer(PropertyReportRenderer renderer) {
     }
 
     @Override
-    public void generate(Project project) {
-        Map<String, Object> entries = new TreeMap<>(project.getProperties());
+    protected PropertyReportModel calculateReportModelFor(Project project) {
+        return computePropertyReportModel(project);
+    }
+
+    private PropertyReportTask.PropertyReportModel computePropertyReportModel(Project project) {
+        PropertyReportModel model = new PropertyReportModel();
+        Map<String, ?> projectProperties = project.getProperties();
         if (property.isPresent()) {
-            String propertyValue = property.get();
-            if (propertyValue.equals("properties")) {
-                renderer.addProperty(propertyValue, "{...}");
+            String propertyName = property.get();
+            if ("properties".equals(propertyName)) {
+                model.putProperty(propertyName, "{...}");
             } else {
-                renderer.addProperty(propertyValue, entries.get(propertyValue));
+                model.putProperty(propertyName, projectProperties.get(propertyName));
             }
         } else {
-            for (Map.Entry<String, ?> entry : entries.entrySet()) {
-                if (entry.getKey().equals("properties")) {
-                    renderer.addProperty(entry.getKey(), "{...}");
+            for (Map.Entry<String, ?> entry : projectProperties.entrySet()) {
+                if ("properties".equals(entry.getKey())) {
+                    model.putProperty(entry.getKey(), "{...}");
                 } else {
-                    renderer.addProperty(entry.getKey(), entry.getValue());
+                    model.putProperty(entry.getKey(), entry.getValue());
                 }
             }
         }
+        return model;
+    }
+
+    @Override
+    protected void generateReportFor(ProjectDetails project, PropertyReportModel model) {
+        for (PropertyWarning warning : model.warnings) {
+            getLogger().warn(
+                "Rendering of the property '{}' with value type '{}' failed with exception",
+                warning.name, warning.valueClass, warning.exception
+            );
+        }
+        for (Map.Entry<String, String> entry : model.properties.entrySet()) {
+            renderer.addProperty(entry.getKey(), entry.getValue());
+        }
+    }
+
+    /**
+     * Model for the report.
+     *
+     * @since 7.6
+     */
+    @Incubating
+    public static final class PropertyReportModel {
+
+        private PropertyReportModel() {
+        }
+
+        private final Map<String, String> properties = new TreeMap<>();
```

Comment:
Perhaps it would be better to sort when building the model and just keep a `List<Pair<String, String>>` in the model.